frontend/src/store/modules/user.ts

```diff
@@ -52,31 +60,55 @@ export default {
     setUserRoles: (state: any, roles: { [key: string]: any } | null) => {
       state.roles = roles;
     },
+    setMetadataCallback: (
+      state: any,
+      callback: ((snapshot: firebase.database.DataSnapshot) => void) | null
+    ) => {
+      state.metadataCallback = callback;
+    },
+    setTimestamp: (state: any, timestamp: number) => {
+      state.timestamp = timestamp;
+    }
   },
   actions: {
     signOut() {
       firebase.auth().signOut();
     },
-    async handleUser(
-      { commit, dispatch }: ActionContext<any, any>,
+    async onAuthStateChanged(
+      { dispatch, commit, state, getters }: ActionContext<any, any>,
       user: firebase.User | null
     ) {
       commit("setCurrentUser", user);
-      if (user) {
-        await dispatch(
-          "updateUserRoles",
-          (await user.getIdTokenResult()).claims.roles
-        );
-      } else {
-        await dispatch("updateUserRoles", null);
+
+      if (state.metadataCallback && getters.metadataRef) {
+        commit("setTimestamp", null);
+        getters.metadataRef.off("value", state.metadataCallback);
       }
+      if (!state.currentUser) {
+        commit("setMetadataCallback", null);
+        commit("setTimestamp", null);
+        await dispatch("updateUserRoles");
+        return;
+      }
+      const callback = () => {
+        if (!state.timestamp) {
+          commit("setTimestamp", +new Date());
```

Comment:
I'm afraid the logic here is not correct. We should save the `refreshTime` that we got from the database. By the way, it's a good candidate for the state property name: `metadataRefreshTime`.

Also, it might be useful to compare the refresh time from the database against the local last refresh time. And avoid updating the roles if the local refresh time is higher.